### DISCUSSION TITLE
Thread exception in playback

### DISCUSSION THREAD

**Discussion by Nickonian** [PaulleDemon/tkVideoPlayer#53] [Q&A]
Hi Paul,

I'm having an issue with the run calls in my python code. My script is a basic GUI to scan for file duplicates and perform actions on the results.  I have a tkVideoPlayer widget in my GUI, which automatically plays a video if a file is selected from a list. The videos play ok, but when switching to a new file I eventually get this exception:

Exception in thread Thread-5 (_load):
Traceback (most recent call last):
  File "/usr/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.10/threading.py", line 953, in run
    self._target(*self._args, **self._kwargs)
  File "/home/<user>/PycharmProjects/pythonProject2/.venv/lib/python3.10/site-packages/tkVideoPlayer/tkvideoplayer.py", line 206, in _load
    self._container.close()
AttributeError: 'NoneType' object has no attribute 'close'

I've traced it through the tkVideoPlayer.py and threading.py to an issue with run calls and the "self._container" value, which occasionally returns as None - here's a debug print I added to your code to show the value of the container. 

container: <av.InputContainer '/home/<user>/Downloads/AATEST/SampleVideo_1280x720_5mb.mkv'>
container: <av.InputContainer '/home/<user>/Downloads/AATEST/SampleVideo_1280x720_5mb(1).mkv'>
container: None
(often followed by PDB hang or exception)

I initially thought the issue may be due to libavcodec issues, but the exception can occur between duplicate files of the same size and codec. The starting point for my debug is obviously my code and the logic of the functions/threads, but I am still in the "learning" phase so my understanding of threading is limited. I'm using Pycharm 3.4 on Ubuntu 22.04.4

Any pointers would be most appreciated!

Cheers, Nick

**Comment by PaulleDemon** [comment]
@Nickonian Did you use `stop()` method before loading the new file?

**Reply by Nickonian** [reply]
Thanks for your reply. I have included a stop function but I'm not sure how elegant/efficient it is!

    def load_and_play_video_thread(self, video_path):
        try:
            # Acquire the lock before starting the thread
            self.thread_lock.acquire()

            # Stop the video playback if it's already playing
            self.stop_video()

            # Create and start the playback thread
            self.playback_thread = threading.Thread(target=self.load_and_play_video, args=(video_path,))
            self.playback_thread.start()
        except Exception as e:
            print(f"Error loading and playing video: {e}")
        finally:
            # Release the lock after the thread finishes
            self.thread_lock.release()
            
         def load_and_play_video(self, video_path):
        """Load and play the video while preserving aspect ratio."""
        # pdb.set_trace()
        try:
            # Set keep_aspect to True
            self.player.keep_aspect(True)

            # Load the video
            self.player.load(video_path)

            # Start playing the video
            self.player.play()

        except Exception as e:
            print(f"Error loading or playing video: {e}")

    def stop_video(self):
        """Stop the video playback."""
        if self.playback_thread and self.playback_thread.is_alive():
            # If the playback thread is still running, join it to wait for completion
            self.playback_thread.join()

            # Ensure the video player stops after the thread finishes
            if self.player:
                self.player.stop()

**Reply by PaulleDemon** [reply]
@Nickonian  It seems like you are running GUI elements from a different thread, all the GUI elements are supposed to be run in the main thread only.
